### PR TITLE
Updated nix overlay to support wayland and X11 at the same time (and remove eww-wayland overlay)

### LINKED ISSUE #733: [BUG] Cannot build .#eww with Nix
### Checklist before submitting an issue

- [X] I have searched through the existing [closed and open issues](https://github.com/elkowar/eww/issues?q=is%3Aissue) for eww and made sure this is not a duplicate
- [X] I have specifically verified that this bug is not a common [user error](https://github.com/elkowar/eww/issues?q=is%3Aissue+label%3Ano-actual-bug+is%3Aclosed)
- [X] I am providing as much relevant information as I am able to in this bug report (Minimal config to reproduce the issue for example, if applicable)

### Description of the bug

Eww cannot be build with nix after changes made on #722.

### Reproducing the issue

```
nix build 'github:elkowar/eww/642983a066931225aef43d4fba77c6d99e8d580d#eww'
error: builder for '/nix/store/kj68pkc9jrkh0v8xw4rzcdg509bg8ps5-eww-642983a066931225aef43d4fba77c6d99e8d580d.drv' failed with exit code 101;
       last 10 log lines:
       >   Package gtk-layer-shell-0 was not found in the pkg-config search path.
       >   Perhaps you should add the directory containing `gtk-layer-shell-0.pc'
       >   to the PKG_CONFIG_PATH environment variable
       >   No package 'gtk-layer-shell-0' found
       >   Package gtk-layer-shell-0 was not found in the pkg-config search path.
       >   Perhaps you should add the directory containing `gtk-layer-shell-0.pc'
       >   to the PKG_CONFIG_PATH environment variable
       >   No package 'gtk-layer-shell-0' found
       >
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/kj68pkc9jrkh0v8xw4rzcdg509bg8ps5-eww-642983a066931225aef43d4fba77c6d99e8d580d.drv'.
```
### Expected behaviour

Running `github:elkowar/eww#eww` should build successfully

### Additional context

```
~/git/eww master
❯ git bisect start                                        
status: waiting for both good and bad commits

~/git/eww master bisect
❯ git bisect good c72b881c3f8f2d2ad8d7a6c337ab3f55ff73ef4b
status: waiting for bad commit, 1 good commit known

~/git/eww master bisect
❯ git bisect bad e76206817de1cb86ec431dcff7d4b04c8b7d36fc
Bisecting: 2 revisions left to test after this (roughly 1 step)
[642983a066931225aef43d4fba77c6d99e8d580d] Default to building with both x11 and wayland (#722)

~/git/eww remotes/origin/fix-dangling-daemon~3 bisect
❯ git bisect run nix build .\#eww                        
running  'nix' 'build' '.#eww'
error: builder for '/nix/store/kj68pkc9jrkh0v8xw4rzcdg509bg8ps5-eww-642983a066931225aef43d4fba77c6d99e8d580d.drv' failed with exit code 101;
       last 10 log lines:
       >   Package gtk-layer-shell-0 was not found in the pkg-config search path.
       >   Perhaps you should add the directory containing `gtk-layer-shell-0.pc'
       >   to the PKG_CONFIG_PATH environment variable
       >   No package 'gtk-layer-shell-0' found
       >   Package gtk-layer-shell-0 was not found in the pkg-config search path.
       >   Perhaps you should add the directory containing `gtk-layer-shell-0.pc'
       >   to the PKG_CONFIG_PATH environment variable
       >   No package 'gtk-layer-shell-0' found
       >
       > warning: build failed, waiting for other jobs to finish...
       For full logs, run 'nix log /nix/store/kj68pkc9jrkh0v8xw4rzcdg509bg8ps5-eww-642983a066931225aef43d4fba77c6d99e8d580d.drv'.
Bisecting: 0 revisions left to test after this (roughly 0 steps)
[de232de41b428b2caf1f41e021e67c0b86e65908] feat: `truncate-left` for labels (#721)
running  'nix' 'build' '.#eww'
642983a066931225aef43d4fba77c6d99e8d580d is the first bad commit
commit 642983a066931225aef43d4fba77c6d99e8d580d
Author: ElKowar <5300871+elkowar@users.noreply.github.com>
Date:   Sun Mar 26 12:09:03 2023 +0200

    Default to building with both x11 and wayland (#722)
    
    * Default to building with both x11 and wayland
    
    * Update CI to include tests and building for wayland and x11

 .github/workflows/build.yml                        |  12 +-
 CHANGELOG.md                                       |   1 +
 Cargo.lock                                         | 348 ++++++++++-----------
 crates/eww/Cargo.toml                              |  37 ++-
 crates/eww/src/app.rs                              |  18 +-
 crates/eww/src/display_backend.rs                  | 197 ++++++------
 crates/eww/src/main.rs                             | 178 ++++++-----
 crates/eww/src/opts.rs                             |   9 +-
 crates/eww/src/server.rs                           |  12 +-
 crates/yuck/Cargo.toml                             |   5 -
 crates/yuck/src/config/backend_window_options.rs   | 193 ++++++------
 .../snapshots/yuck__parser__test__test-10.snap     |   8 +
 .../snapshots/yuck__parser__test__test-11.snap     |   8 +
 .../snapshots/yuck__parser__test__test-12.snap     |   8 +
 .../snapshots/yuck__parser__test__test-13.snap     |   8 +
 .../snapshots/yuck__parser__test__test-14.snap     |   8 +
 .../snapshots/yuck__parser__test__test-15.snap     |   8 +
 .../snapshots/yuck__parser__test__test-16.snap     |   8 +
 .../snapshots/yuck__parser__test__test-17.snap     |   8 +
 .../snapshots/yuck__parser__test__test-2.snap      |   8 +
 .../snapshots/yuck__parser__test__test-3.snap      |   8 +
 .../snapshots/yuck__parser__test__test-4.snap      |   8 +
 .../snapshots/yuck__parser__test__test-5.snap      |   8 +
 .../snapshots/yuck__parser__test__test-6.snap      |   8 +
 .../snapshots/yuck__parser__test__test-7.snap      |   8 +
 .../snapshots/yuck__parser__test__test-8.snap      |   8 +
 .../snapshots/yuck__parser__test__test-9.snap      |   8 +
 .../parser/snapshots/yuck__parser__test__test.snap |   8 +
 28 files changed, 658 insertions(+), 488 deletions(-)
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-10.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-11.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-12.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-13.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-14.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-15.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-16.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-17.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-2.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-3.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-4.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-5.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-6.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-7.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-8.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test-9.snap
 create mode 100644 crates/yuck/src/parser/snapshots/yuck__parser__test__test.snap
bisect found first bad commit
```

### PR DESCRIPTION
**Pull request by Philipp-M**
## Description

Fixes #733

Since X11 and wayland is now supported at the same time, eww could be reduced to the same package `eww`.
Right now this builds with support for X11 and wayland (default features). It may make sense to add more fine-grained support for different features in upstream via parameters, when it is updated there (i.e. `withWayland`, `withX11`, for either X11 or wayland support or both at the same time (should probably be the default then)) .

### CODE REVIEW

**Comment by adomixaszvers** [comment]
As of 82e13c18d3bbba3ab9a2bff8096a2ae98b030b35, it builds only with wayland feature so it opens as a regular window on x11.

**Comment by adomixaszvers** [comment]
It does not build with x11 feature.
```
$ nix build --no-link 'github:Philipp-M/eww/df6bed4d6aee994dac231a787b11fde900a2c486#eww'

$ nix log 'github:Philipp-M/eww/df6bed4d6aee994dac231a787b11fde900a2c486#eww' 
```

Cargo invokation was
```
env CC_x86_64-unknown-linux-gnu=/nix/store/3d9zjv5vaqjxb9kwbdqsd194alwm97x1-gcc-wrapper-11.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/3d9zjv5vaqjxb9kwbdqsd194alwm97x1-gcc-wrapper-11.3.0/bin/c++ CC_x86_64-unknown-linux-gnu=/nix/store/3d9zjv5vaqjxb9kwbdqsd194alwm97x1-gcc-wrapper-11.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/3d9zjv5vaqjxb9kwbdqsd194alwm97x1-gcc-wrapper-11.3.0/bin/c++ cargo build -j 8 --target x86_64-unknown-linux-gnu --frozen --release --features=wayland --bin eww
```

**Comment by Philipp-M** [comment]
Are you sure, that it's not having the x11 feature? AFAIK features are additive as long as default features aren't explicitly disabled. So this should be equivalent to just (roughly) `cargo build`.

File: flake.nix [hunk 1]
```diff
@@ -30,15 +30,18 @@
...
         in
-        {
-          eww = (prev.eww.override { inherit rustPlatform; }).overrideAttrs (old: {
+        rec {
+          # TODO update this to reflect the changes in upstream (nixpkgs), when upstream is updated
+          eww = (prev.eww.override { inherit rustPlatform; withWayland = true; }).overrideAttrs (old: {
             version = self.rev or "dirty";
             src = builtins.path { name = "eww"; path = prev.lib.cleanSource ./.; };
             cargoDeps = rustPlatform.importCargoLock { lockFile = ./Cargo.lock; };
+            cargoBuildNoDefaultFeatures = false;
```
Comment (adomixaszvers) [inline flake.nix:39]: Double negative (`cargoBuildNoDefaultFeatures` and `false`) is confusing a bit.

Comment (Philipp-M) [reply flake.nix:39]: Yeah I also don't understand why the authors of `buildRustPackage` have chosen the naming of this...
But AFAIK it's the var that's necessary to use to enable/disable default features

**Review by adomixaszvers — APPROVED** [review]

File: flake.nix [hunk 2]
```diff
@@ -6,7 +6,7 @@
     rust-overlay.inputs.nixpkgs.follows = "nixpkgs";
   };
 
-  outputs = { self, nixpkgs, rust-overlay, flake-compat, ... }:
```
Comment (fortuneteller2k) [inline flake.nix:9]: `flake-compat` is used in default.nix, revert this change

**Review by fortuneteller2k — CHANGES_REQUESTED** [review]

File: flake.nix (hunk 2 above)
Comment (Philipp-M) [reply flake.nix:9]: This should be a non issue though, as `default.nix` reads directly from the `flake.lock` and this is just a destructuring pattern of the function argument (`flake-compat` is not used anywhere in the `flake.nix`). I have not removed the flake input though (which AFAIK is the only necessary thing for an entry in the `flake.lock`).

File: flake.nix (hunk 1 above)
Comment (eclairevoyant) [reply flake.nix:39]: It's probably due to how cargo named it (`cargo build --no-default-features`)

**Comment by eclairevoyant** [comment]
Was this tested (runtime)? I imagine #739 would cause problems and this change maybe doesn't make sense to merge as is (right now)

fwiw turning off default features and building with wayland does work regardless

**Comment by Philipp-M** [comment]
You're correct just tested it in hyprland, seems like I forgot to test this properly with wayland, #739 seems to be an issue. Well than the question arises, if #739 should be fixed first, or I'm handling the cases differently (i.e. disable the `x11` feature for wayland support). I believe fixing #739 would be better to be able to use one binary for all backends.

I think it makes sense to offer multiple flake outputs (`eww-x11`, `eww-wayland` and `eww` which supports both backends)

**Comment by eclairevoyant** [comment]
> I believe fixing #739 would be better to be able to use one binary for all backends.

Of course, I agree, my suggestion to build wayland binary without x11 support was just a workaround until #739 is addressed (if this PR were to be merged first). Other option would be to hold off on this PR for now

**Comment by Philipp-M** [comment]
Since it's a simple change, I've updated the PR to support wayland separately (by implicitly disabling the feature `x11` (done by nixpkgs upstream). That way it should be ready to merge anyways (the output `eww` just doesn't support wayland currently due to #739)

**Comment by adomixaszvers** [comment]
https://github.com/NixOS/nixpkgs/pull/239826

**Comment by w-lfchen** [comment]
The reliance on nixpkgs has been removed, so these changes seem obsolete to me

**Comment by Philipp-M** [comment]
Yeah, I'll close it, it's very likely obsolete or completely out of date.
